src/pip/_internal/locations/_sysconfig.py

```diff
@@ -130,6 +155,12 @@ def get_scheme(
     else:
         scheme_name = _infer_prefix()
 
+    # Special case: When installing into a custom prefix, use posix_prefix
+    # instead of osx_framework_library. See _should_use_osx_framework_prefix()
+    # docstring for details.
+    if prefix is not None and scheme_name == "osx_framework_library":
```

Comment:
If my understanding is correct, this can only correspond to the situation where we've gone into the `scheme_name = _infer_prefix()` case and that particular function is only called from here.

Hence, I wonder if it'd make sense to have that particular piece of logic directly in the  `_infer_prefix` rather than as step performed afterward.

This could correspond to doing something like:

- adding a param to the _infer_prefix function about the prefix being provided or not
- writing `return "osx_framework_library" if provided_prefix is None else "posix_prefix"` in the _infer_prefix function

Reply:
My intention behind the current approach is to make those`_infer_*` functions act similarly to `sysconfig.get_preferred_scheme` (because eventually we’re going to replace the ad-hoc inference with it). If we put the logic in `_infer_prefix`, when we remove `_infer_user` and `_infer_home`, we’ll still be left with one short `_infer_prefix` function. I don’t think either structure is better though, this is just how I mentally structure this logic (first I ask the macOS Python what its preferred scheme is, and if it says `osx_framework_library`, oh I’m going to ignore it).